spring-integration-file/src/main/java/org/springframework/integration/file/remote/session/DelegatingSessionFactory.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.file.remote.session;
+
+import java.util.Map;
+
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+
+/**
+ * {@link SessionFactory} that delegates to a {@link SessionFactory} retrieved from a
+ * {@link SessionFactoryFactory}.
+ *
+ * @author Gary Russell
+ * @since 4.2
+ *
+ */
+public class DelegatingSessionFactory<F> implements SessionFactory<F> {
+
+	private final SessionFactoryFactory<F> factoryFactory;
+
+	private final ThreadLocal<Object> threadKey = new ThreadLocal<Object>();
```

Comment:
The "strategy" here is `SessionFactory<F>` - that's why I abstracted  `SessionFactoryLocator`.

This is very lightweight - if someone wants another implementation then replace this only.

Maybe this should be called `ThreadLocalDelegatingSessionFactory` or similar.